src/inbox.rs

```diff
@@ -0,0 +1,283 @@
+// TODO(bounded)
+
+mod rx;
+mod tx;
+
+pub(crate) use rx::Receiver;
+pub(crate) use tx::{SendFuture, Sender};
+
+use crate::envelope::{BroadcastEnvelope, MessageEnvelope};
+
+use crate::inbox::tx::WaitingSender;
+use std::cmp::Ordering;
+use std::collections::{BinaryHeap, VecDeque};
+
+use std::sync::{Arc, Mutex, Weak};
+
+use crate::inbox::rx::WaitingReceiver;
+
+type Chan<A> = Arc<Mutex<Inner<A>>>;
+type Spinlock<T> = spin::Mutex<T>;
+type StolenMessage<A> = Box<dyn MessageEnvelope<Actor = A>>;
+type BroadcastMessage<A> = Arc<dyn BroadcastEnvelope<Actor = A>>;
+type BroadcastQueue<A> = Spinlock<BinaryHeap<BroadcastMessageWrapper<A>>>;
+
+// TODO(priority)
+#[derive(PartialEq, Eq, Ord, PartialOrd, Copy, Clone)]
+pub(crate) enum Priority {
+    Min,
+    Valued(i32),
+}
+
+impl Default for Priority {
+    fn default() -> Self {
+        Priority::Valued(0)
+    }
+}
+
+pub(crate) fn new<A>(bound: Option<usize>) -> (Sender<A>, Receiver<A>) {
+    let broadcast_mailbox = Arc::new(Spinlock::new(BinaryHeap::new()));
+    let inner = Inner {
+        default_queue: VecDeque::new(),
+        bound,
+        waiting_senders: VecDeque::new(),
+        waiting_receivers: VecDeque::new(),
+        priority_queue: BinaryHeap::new(),
+        broadcast_queues: vec![Arc::downgrade(&broadcast_mailbox)],
+        shutdown: false,
+    };
+
+    let inner = Arc::new(Mutex::new(inner));
+
+    let rx = Receiver {
+        inner: inner.clone(),
+        broadcast_mailbox,
+    };
+
+    (Sender(inner), rx)
+}
+
+// TODO(perf): try and reduce contention on Inner as much as possible
+// Might be able to move some stuff out to atomics, or lock it separately. This should net some
+// overall performance gains, but these likely won't in crude_bench or any throughput testing
+struct Inner<A> {
+    default_queue: VecDeque<StolenMessage<A>>,
+    bound: Option<usize>,
+    waiting_senders: VecDeque<Weak<Spinlock<WaitingSender<A>>>>,
+    waiting_receivers: VecDeque<Weak<Spinlock<WaitingReceiver<A>>>>,
+    priority_queue: BinaryHeap<StolenMessageWithPriority<A>>,
+    broadcast_queues: Vec<Weak<BroadcastQueue<A>>>,
+    shutdown: bool,
+}
+
+impl<A> Inner<A> {
+    fn is_full(&self) -> bool {
+        self.bound
+            .map(|cap| self.default_queue.len() >= cap)
+            .unwrap_or(false)
+    }
+
+    fn pop_priority(&mut self) -> Option<StolenMessage<A>> {
+        self.priority_queue.pop().map(|msg| msg.val)
+    }
+
+    fn pop_default(&mut self) -> Option<StolenMessage<A>> {
+        // If len < cap after popping this message, try fulfill at most one waiting sender
+        if self.is_full() {
+            self.try_fulfill_sender();
+        }
+
+        self.default_queue.pop_front()
+    }
+
+    fn try_fulfill_receiver(&mut self, mut reason: WakeReason<A>) -> Result<(), WakeReason<A>> {
+        while !self.waiting_receivers.is_empty() {
+            if let Some(rx) = self.waiting_receivers.pop_front().and_then(|w| w.upgrade()) {
```

Comment:
I think so. At some point it was a while let, and then I changed it back maybe because of borrowing stuff but that doesn't appear to be relevant anymore

Reply:
Actually, I'm not so sure. This would lead to it ending as soon as one receiver encountered was a `Weak` that couldn't be upgraded. Might be possible as an iterator, though

Reply:
That could work